### ISSUE TITLE
Holes in mesh with CWBVH

### ISSUE THREAD

**Issue by andr3wmac** [jbikker/tinybvh#28]
When I try more dense geometry I'm starting to see significant holes in the meshes but only when the intersection is set to CWBVH. 

Its harder to notice with Sponza but the lions heads are missing some triangles, using happy buddha the problem becomes very pronounced. I've attached two screenshots that use Intersect, one set to WALD_32BYTE and one set to CWBVH. 

![CPUTrace_WALD_32BYTE](https://github.com/user-attachments/assets/50c842fc-cc78-42bd-9733-95b5ddea1dd7)
![CPUTrace_CWBVH](https://github.com/user-attachments/assets/db0c227f-8b76-4796-97fa-22eed6161437)

**Comment by jbikker** [comment]
Ah yes, I forgot about that The CWBVH format has a strict limitation for leaf primitive count: It can only be 1, 2 or 3. For Sponza this didn't show up as an issue, but in your scene it clearly does. 
What you can try, pending a good solution: Building the BVH can already be done 'as far as possible', i.e. ignoring stop signals of the SAH heuristic. If even that isn't sufficient, you can use 'SplitLeafs' to reduce each leaf to a single primitive. To do that: Convert to verbose, SplitLeafs, then convert to CWBVH *via WALD_32BYTE*.
A proper solution should build the BVH with deep splits, then merge (MergeLeafs) with a limit on leaf prim count (not implemented yet).

**Comment by andr3wmac** [comment]
Can confirm splitting leafs into single triangles resolves the issue. I assume SplitLeafs(3) is a pending change you have limit it to 3 per leaf which sounds better than going to single triangles. I'll leave it up to you on when to close this issue as my issue is technically resolved but someone else trying CWBVH will likely still run into the same thing.

**Comment by jbikker** [comment]
I'll leave it open for a bit as this is an annoying limitation to CWBVH. I will either make it split while converting to CWBVH or remove the restriction somehow: It seems rather wasteful. SplitLeafs is not really suitable here; this really only serves to feed Optimize.

**Comment by jbikker** [comment]
OK, this uncovered a serious bug in all builders, which was introduced with the recent C_INT / C_TRAV feature: With a non-zero C_TRAV, SAH steered subdiv termination failed, causing massive leafs (133 prims and more). The weird CWBVH limitation made this painfully obvious. This has now beemn resolved, but CWBVH still needs to be modified to handle cases where 4 or more prims happen. So I'll still leave this open. You won't need the kludge anymore for Buddha though.

**Comment by jbikker** [comment]
Final message on this: With the fix to the builders, leafs with more than 3 primitives are now very rare: It never happens in 'happy buddha', and there are only two instances in Crytek Sponza (one 6-prim leaf, one 4-prim leaf). These cases are now handled by splitting the leaf during conversion, which means that pre-splitting is no longer needed and any BVH can directly be converted to CWBVH. That should fully fix the issue.

**Comment by andr3wmac** [comment]
I updated to the latest and removed the SplitLeafs and extra conversions and it looked good at first with buddha but then I noticed missing triangles in Sponza. SplitLeafs(1) and SplitLeafs(3) produce the same result which is correct however going from WALD_32BYTE -> BASIC_BVH8 -> CWBVH without VERBOSE/SplitLeafs causes holes. I've attached two screenshots below of SplitLeafs followed by no SplitLeafs.

![image](https://github.com/user-attachments/assets/b2a6ba5f-93b0-44a2-9710-e59fd1fd9caa)
![image](https://github.com/user-attachments/assets/be60fad4-bad7-45ce-8141-9001633e8d8f)
